File: backend/universe.py
"""0-fee symbol universe manager.

Periodically pulls the list of 0-fee MEXC futures symbols, resolves available
reference symbols on Binance Futures, and exposes the working set.
"""

from __future__ import annotations

import asyncio
import json
import logging
import os
import time
from dataclasses import dataclass
from pathlib import Path
from typing import Dict, List, Optional, Set

import aiohttp


logger = logging.getLogger(__name__)


BINANCE_FUTURES_INFO = "https://fapi.binance.com/fapi/v1/exchangeInfo"


def to_binance(mexc: str) -> str:
    """BTC_USDT -> BTCUSDT."""
    return mexc.replace("_", "").upper()


def resolve_binance_symbol(mexc: str, binance_set: set) -> tuple:
    """Try multiple Binance symbol variants for a given MEXC symbol.

    Returns (binance_symbol, price_factor) where price_factor is the multiplier
    to convert Binance price → MEXC-equivalent price.
    e.g. 1000PEPEUSDT price × 0.001 = PEPE_USDT fair price.
    """
    base = to_binance(mexc)
    if base in binance_set:
        return base, 1.0
    # Try 1000-prefix variant (common for micro-cap coins)
    prefixed = f"1000{base}"
    if prefixed in binance_set:
        return prefixed, 0.001
    return None, 1.0


def to_mexc(binance: str) -> str:
    """BTCUSDT -> BTC_USDT (best-effort, USDT only)."""
    s = binance.upper()
    if s.endswith("USDT") and "_" not in s:
        return f"{s[:-4]}_USDT"
    return s


@dataclass
class UniverseEntry:
    mexc_symbol: str
    binance_symbol: Optional[str]
    has_ref: bool
    binance_price_factor: float = 1.0  # multiply Binance price by this to get MEXC-equivalent fair


class UniverseManager:
    def __init__(
        self,
        mexc_trader,
        cfg,
        cache_path: Optional[Path] = None,
    ) -> None:
        self.trader = mexc_trader
        self.cfg = cfg
        self.cache_path = cache_path or Path(
            os.environ.get("ZFEE_UNIVERSE_CACHE")
            or (Path(__file__).resolve().parents[1] / ".universe_cache.json")
        )

        self._binance_set: Set[str] = set()
        self._binance_set_ts: float = 0.0
        self._entries: Dict[str, UniverseEntry] = {}
        self._lock = asyncio.Lock()

    @property
    def entries(self) -> Dict[str, UniverseEntry]:
        return self._entries

    @property
    def working_set(self) -> List[str]:
        """MEXC symbols we should subscribe to."""
        out: List[str] = []
        excl = set(s.upper() for s in (self.cfg.universe.exclude or []))
        only = set(s.upper() for s in (self.cfg.universe.include_only or []))
        for sym, e in self._entries.items():
            if sym.upper() in excl:
                continue
            if only and sym.upper() not in only:
                continue
            if self.cfg.universe.require_binance_ref and not e.has_ref:
                continue
            out.append(sym)
            if len(out) >= self.cfg.universe.max_symbols:
                break
        return out

    @property
    def available_pool(self) -> List[str]:
        """All MEXC 0-fee symbols that also have a Binance reference,
        regardless of include_only / exclude / max_symbols."""
        out: List[str] = []
        for sym, e in self._entries.items():
            if self.cfg.universe.require_binance_ref and not e.has_ref:
                continue
            out.append(sym)
        return out

    def reference_for(self, mexc_symbol: str) -> Optional[str]:
        e = self._entries.get(mexc_symbol)
        return e.binance_symbol if e else None

    def price_factor_for(self, mexc_symbol: str) -> float:
        e = self._entries.get(mexc_symbol)
        return e.binance_price_factor if e else 1.0

    async def refresh(self) -> List[str]:
        """Pull 0-fee list + Binance availability. Returns the working set."""
        async with self._lock:
            zero_fee = await self._safe_zero_fee_list()
            binance_set = await self._safe_binance_set()

            entries: Dict[str, UniverseEntry] = {}

            # Force-include symbols come first so they're not pushed off by max_symbols cap.
            for sym in (self.cfg.universe.force_include_symbols or []):
                norm = str(sym).strip().upper()
                if not norm:
                    continue
                if not norm.endswith("_USDT") and "_" not in norm:
                    norm = f"{norm}_USDT"
                if norm in entries:
                    continue
                bsym, bfactor = resolve_binance_symbol(norm, binance_set)
                has_ref = bsym is not None
                entries[norm] = UniverseEntry(
                    mexc_symbol=norm,
                    binance_symbol=bsym,
                    has_ref=has_ref,
                    binance_price_factor=bfactor,
                )

            for sym in zero_fee:
                if sym in entries:
                    continue
                bsym, bfactor = resolve_binance_symbol(sym, binance_set)
                has_ref = bsym is not None
                entries[sym] = UniverseEntry(
                    mexc_symbol=sym,
                    binance_symbol=bsym,
                    has_ref=has_ref,
                    binance_price_factor=bfactor,
                )

            self._entries = entries
            try:
                self._save_cache()
            except Exception:
                pass
            return self.working_set

    async def loop(self, on_change=None) -> None:
        """Long-running refresh loop."""
        while True:
            try:
                prev = set(self.working_set)
                ws = await self.refresh()
                cur = set(ws)
                if prev != cur and on_change is not None:
                    try:
                        await on_change(ws)
                    except Exception as e:
                        logger.warning("universe on_change handler error: %s", e)
            except Exception as e:
                logger.warning("universe refresh error: %s", e)
            await asyncio.sleep(max(30, int(self.cfg.universe.refresh_sec)))

    # -------------------- internals --------------------

    async def _safe_zero_fee_list(self) -> List[str]:
        try:
            return await self.trader.list_zero_fee_symbols()
        except Exception as e:
            logger.warning("list_zero_fee_symbols failed: %s; using cache", e)
            return self._load_cache().get("symbols", [])

    async def _safe_binance_set(self) -> Set[str]:
        # Cache for 10 minutes
        now = time.time()
        if self._binance_set and (now - self._binance_set_ts) < 600:
            return self._binance_set

        try:
            connector = aiohttp.TCPConnector(ssl=False, resolver=aiohttp.ThreadedResolver())
            async with aiohttp.ClientSession(
                connector=connector,
                timeout=aiohttp.ClientTimeout(total=10)
            ) as s:
                async with s.get(BINANCE_FUTURES_INFO) as r:
                    data = await r.json()
            symbols = data.get("symbols") or []
            out = set()
            for it in symbols:
                if not isinstance(it, dict):
                    continue
                sym = str(it.get("symbol") or "").upper()
                contract_type = str(it.get("contractType") or "").upper()
                quote = str(it.get("quoteAsset") or "").upper()
                if quote != "USDT":
                    continue
                if contract_type and contract_type != "PERPETUAL":
                    continue
                out.add(sym)
            self._binance_set = out
            self._binance_set_ts = now
            return out
        except Exception as e:
            logger.warning("Binance exchangeInfo failed: %s", e)
            return self._binance_set or set()

    def _save_cache(self) -> None:
        data = {
            "ts": time.time(),
            "symbols": list(self._entries.keys()),
            "with_ref": [s for s, e in self._entries.items() if e.has_ref],
        }
        self.cache_path.write_text(
            json.dumps(data, ensure_ascii=False), encoding="utf-8"
        )

    def _load_cache(self) -> dict:
        try:
            return json.loads(self.cache_path.read_text(encoding="utf-8"))
        except Exception:
            return {}

File: rust-bot/src/universe.rs
use crate::config::{AppConfig, normalize_symbol_name};
use crate::mexc::MexcClient;
use crate::state::now_ts;
use serde::{Deserialize, Serialize};
use serde_json::json;
use std::collections::{HashMap, HashSet};
use std::fs;
use std::path::PathBuf;

const BINANCE_FUTURES_INFO: &str = "https://fapi.binance.com/fapi/v1/exchangeInfo";

pub fn to_binance(mexc: &str) -> String {
    mexc.replace('_', "").to_ascii_uppercase()
}

fn price_factor_for_variant(mexc: &str, binance_symbol: &str) -> f64 {
    let base = to_binance(mexc);
    if binance_symbol.is_empty() || binance_symbol == base {
        return 1.0;
    }
    if let Some(prefix) = binance_symbol.strip_suffix(&base) {
        if let Ok(mult) = prefix.parse::<f64>() {
            if mult > 0.0 {
                return 1.0 / mult;
            }
        }
    }
    1.0
}

pub fn resolve_binance_symbol(
    mexc: &str,
    binance_set: &HashSet<String>,
    alias_map: &HashMap<String, String>,
) -> (Option<String>, f64) {
    let mexc_u = mexc.to_ascii_uppercase();
    let base = to_binance(&mexc_u);
    if let Some(alias) = alias_map.get(&mexc_u).map(|s| s.to_ascii_uppercase()) {
        if binance_set.contains(&alias) {
            let factor = price_factor_for_variant(&mexc_u, &alias);
            return (Some(alias), factor);
        }
    }
    if binance_set.contains(&base) {
        return (Some(base), 1.0);
    }
    let prefixed = format!("1000{base}");
    if binance_set.contains(&prefixed) {
        return (Some(prefixed), 0.001);
    }
    (None, 1.0)
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct UniverseEntry {
    pub mexc_symbol: String,
    pub binance_symbol: Option<String>,
    pub has_ref: bool,
    pub binance_price_factor: f64,
}

#[derive(Debug)]
pub struct UniverseManager {
    cfg: AppConfig,
    trader: MexcClient,
    cache_path: PathBuf,
    binance_set: HashSet<String>,
    binance_set_ts: f64,
    entries: HashMap<String, UniverseEntry>,
    client: reqwest::Client,
}

impl UniverseManager {
    pub fn new(trader: MexcClient, cfg: AppConfig, cache_path: PathBuf) -> Self {
        Self {
            cfg,
            trader,
            cache_path,
            binance_set: HashSet::new(),
            binance_set_ts: 0.0,
            entries: HashMap::new(),
            client: reqwest::Client::new(),
        }
    }

    pub fn update_config(&mut self, cfg: AppConfig) {
        self.cfg = cfg;
    }

    pub fn entries(&self) -> HashMap<String, UniverseEntry> {
        self.entries.clone()
    }

    pub fn working_set(&self) -> Vec<String> {
        let excl: HashSet<String> = self
            .cfg
            .universe
            .exclude
            .iter()
            .map(|s| s.to_ascii_uppercase())
            .collect();
        let only: HashSet<String> = self
            .cfg
            .universe
            .include_only
            .iter()
            .map(|s| s.to_ascii_uppercase())
            .collect();
        let mut out = Vec::new();
        for (sym, entry) in &self.entries {
            if excl.contains(&sym.to_ascii_uppercase()) {
                continue;
            }
            if !only.is_empty() && !only.contains(&sym.to_ascii_uppercase()) {
                continue;
            }
            if self.cfg.universe.require_binance_ref && !entry.has_ref {
                continue;
            }
            out.push(sym.clone());
            if out.len() >= self.cfg.universe.max_symbols {
                break;
            }
        }
        out
    }

    pub fn available_pool(&self) -> Vec<String> {
        self.entries
            .iter()
            .filter_map(|(sym, entry)| {
                (!self.cfg.universe.require_binance_ref || entry.has_ref).then_some(sym.clone())
            })
            .collect()
    }

    pub async fn refresh(&mut self) -> Vec<String> {
        let zero_fee = self.safe_zero_fee_list().await;
        let binance_set = self.safe_binance_set().await;
        let alias_map = self.cfg.binance_symbol_overrides.clone();
        let mut entries = HashMap::new();

        for raw in self
            .cfg
            .universe
            .force_include_symbols
            .iter()
            .chain(zero_fee.iter())
        {
            let sym = normalize_symbol_name(raw);
            if sym.is_empty() || entries.contains_key(&sym) {
                continue;
            }
            let (bsym, factor) = resolve_binance_symbol(&sym, &binance_set, &alias_map);
            entries.insert(
                sym.clone(),
                UniverseEntry {
                    mexc_symbol: sym,
                    has_ref: bsym.is_some(),
                    binance_symbol: bsym,
                    binance_price_factor: factor,
                },
            );
        }
        self.entries = entries;
        self.save_cache();
        self.working_set()
    }

    async fn safe_zero_fee_list(&self) -> Vec<String> {
        let live = self.trader.list_zero_fee_symbols().await;
        if !live.is_empty() {
            live
        } else {
            self.load_cache_symbols()
        }
    }

    async fn safe_binance_set(&mut self) -> HashSet<String> {
        let now = now_ts();
        if !self.binance_set.is_empty() && now - self.binance_set_ts < 600.0 {
            return self.binance_set.clone();
        }
        let result = async {
            let data: serde_json::Value = self
                .client
                .get(BINANCE_FUTURES_INFO)
                .send()
                .await?
                .json()
                .await?;
            Ok::<_, anyhow::Error>(data)
        }
        .await;
        let Ok(data) = result else {
            return self.binance_set.clone();
        };
        let mut out = HashSet::new();
        for item in data
            .get("symbols")
            .and_then(|v| v.as_array())
            .cloned()
            .unwrap_or_default()
        {
            let sym = item
                .get("symbol")
                .and_then(|v| v.as_str())
                .unwrap_or("")
                .to_ascii_uppercase();
            let quote = item
                .get("quoteAsset")
                .and_then(|v| v.as_str())
                .unwrap_or("")
                .to_ascii_uppercase();
            let contract_type = item
                .get("contractType")
                .and_then(|v| v.as_str())
                .unwrap_or("")
                .to_ascii_uppercase();
            if quote != "USDT" {
                continue;
            }
            if !contract_type.is_empty()
                && !matches!(contract_type.as_str(), "PERPETUAL" | "TRADIFI_PERPETUAL")
            {
                continue;
            }
            if !sym.is_empty() {
                out.insert(sym);
            }
        }
        if !out.is_empty() {
            self.binance_set = out;
            self.binance_set_ts = now;
        }
        self.binance_set.clone()
    }

    fn save_cache(&self) {
        let data = json!({
            "ts": now_ts(),
            "symbols": self.entries.keys().cloned().collect::<Vec<_>>(),
            "with_ref": self.entries.iter().filter_map(|(s, e)| e.has_ref.then_some(s.clone())).collect::<Vec<_>>(),
        });
        if let Some(parent) = self.cache_path.parent() {
            let _ = fs::create_dir_all(parent);
        }
        let _ = fs::write(&self.cache_path, data.to_string());
    }

    fn load_cache_symbols(&self) -> Vec<String> {
        let Ok(raw) = fs::read_to_string(&self.cache_path) else {
            return Vec::new();
        };
        let Ok(v) = serde_json::from_str::<serde_json::Value>(&raw) else {
            return Vec::new();
        };
        v.get("symbols")
            .and_then(|v| v.as_array())
            .map(|arr| {
                arr.iter()
                    .filter_map(|s| s.as_str().map(ToString::to_string))
                    .collect()
            })
            .unwrap_or_default()
    }
}
